src/cluster/cluster.cc

```diff
@@ -511,37 +552,122 @@ std::string Cluster::GenNodesDescription() {
-    if (n->slots_info_.size() > 0) n->slots_info_.pop_back();  // Trim space
     if (n->role_ == kClusterMaster && n->slots_info_.size() > 0) {
       node_str.append(" " + n->slots_info_);
     }
-    n->slots_info_.clear();  // Reset
+    nodes_info.append(node_str + "\n");
+  }
+  return nodes_info;
+}
 
-    nodes_desc.append(node_str + "\n");
+Status Cluster::DumpClusterNodes(const std::string &file) {
+  // Parse and validate the cluster nodes string before dumping into file
+  std::string tmp_path = file + ".tmp";
+  remove(tmp_path.data());
+  std::ofstream output_file(tmp_path, std::ios::out);
+  output_file << "# version\n";
+  output_file << fmt::format("{}\n", version_);
+  output_file << "# id\n";
+  output_file << fmt::format("{}\n", myid_);
+  output_file << "# nodes\n";
+  output_file << GenNodesInfo();
+  output_file.close();
+  if (rename(tmp_path.data(), file.data()) < 0) {
+    return {Status::NotOK, fmt::format("rename file encounter error: {}", strerror(errno))};
   }
-  return nodes_desc;
+  return Status::OK();
+}
+
+LoadState parseLoadState(const std::string &in) {
+  if (in == "version") {
+    return LoadState::Version;
+  } else if (in == "id") {
+    return LoadState::ID;
+  } else if (in == "nodes") {
+    return LoadState::NodesInfo;
+  } else {
+    return LoadState::Unknown;
+  }
+}
+
+Status Cluster::LoadClusterNodes(const std::string &file_path) {
+  if (rocksdb::Env::Default()->FileExists(file_path).IsNotFound()) {
+    LOG(INFO) << fmt::format("The cluster nodes file {} is not found. Use CLUSTERX subcommands to specify it.",
+                             file_path);
+    return Status::OK();
+  }
+
+  std::ifstream file;
+  file.open(file_path);
+  if (!file.is_open()) {
+    return {Status::NotOK, fmt::format("error opening the file '{}': {}", file_path, strerror(errno))};
+  }
+
+  int64_t version = -1;
+  std::string id, nodesInfo;
+  LoadState state = LoadState::Unknown;
+  std::string line;
+  while (!file.eof()) {
+    std::getline(file, line);
+    line = Util::Trim(line, " \t");
+    if (line.empty()) continue;  // skip empty line
+
+    if (line[0] == '#') {
+      std::string comment = Util::ToLower(Util::Trim(line, "# \t"));
+      state = parseLoadState(comment);
+      continue;
+    }
+
+    switch (state) {
+      case LoadState::Version: {
+        auto parse_result = ParseInt<int64_t>(line, 10);
+        if (!parse_result) {
+          return {Status::NotOK, errInvalidClusterVersion};
+        }
+        version = *parse_result;
+        break;
+      }
+      case LoadState::ID:
+        id = line;
+        if (id.length() != kClusterNodeIdLen) {
+          return {Status::NotOK, errInvalidNodeID};
+        }
+        break;
+      case LoadState::NodesInfo:
+        // std::getline swallowed the newline in the node info, so need to add it back here
+        nodesInfo.append(line + "\n");
+        break;
+      default:
+        return {Status::NotOK, "got unknown parse state"};
```

Comment:
In the future, if we support other states, seems load would fail?

Reply:
It should make sense that the older version can't parse new state.

Reply:
Yes, I have no strong point in this scenario. My initial intention is to prevent unexpected configurations.

Reply:
To see other folks have any advice? @PragmaTwice @ShooterIT @torwig